This lab on Cross-Validation and Bootstrap in R comes from p. 190-197 of "Introduction to Statistical Learning with
Applications in R" by Gareth James, Daniela Witten, Trevor Hastie and Robert Tibshirani. It was re-implemented in Fall 2016 in `tidyverse` format by Amelia McNamara and R. Jordan Crouser at Smith College.

In [1]:
list.of.packages <- c("boot","broom","dplyr","ISLR","modelr","purrr","tidyr")
new.packages <- list.of.packages[!(list.of.packages %in% installed.packages()[,"Package"])]
if(length(new.packages)) install.packages(new.packages)


In [2]:

library(ISLR)
library(dplyr)
library(broom)
library(modelr)
library(purrr)
library(tidyr)


Attaching package: ‘dplyr’

The following objects are masked from ‘package:stats’:

    filter, lag

The following objects are masked from ‘package:base’:

    intersect, setdiff, setequal, union


Attaching package: ‘modelr’

The following object is masked from ‘package:broom’:

    bootstrap



# 5.3.1 The Validation Set Approach

In this section, we'll explore the use of the validation set approach in order to estimate the
test error rates that result from fitting various linear models on the `Auto`data set.

Before we begin, we use the `set.seed()`function in order to set a seed for
`R}$’s random number generator, so that you'll obtain precisely the same results as those shown in the textbook. It is generally a good idea to set a random seed when performing an analysis such as cross-validation
that contains an element of randomness, so that the results obtained can be reproduced precisely at a later time.

We begin by using the `sample_n()` and `setdiff()` functions to split the set of observations into two halves. We'll start by selecting a random subset of 196 observations out of the original 392 observations. We refer to these observations as the training
set.

In [3]:
set.seed(1)

train = Auto %>%
  sample_n(196)

test = Auto %>%
  setdiff(train)

We then use `lm()`to fit a linear regression using only
the observations corresponding to the training set.

In [4]:
model_LR = lm(mpg~horsepower, data=train)

We now use the `predict()`function to estimate the response for the test
observations, and we use the `mean()`function to calculate the MSE of the
196 observations in the test set:

In [5]:
test = test %>%
  mutate(predictions_quad = predict(model_LR, test))

test %>%
  summarize(MSE_slr = mean((mpg-predictions_quad)^2))

MSE_slr 
1 26.14142

Therefore, the estimated test MSE for the linear regression fit is 26.14. We
can use the `poly()`function to estimate the test error for the quadratic
and cubic regressions.

In [6]:
model_QUAD = lm(mpg~poly(horsepower,2), data=train)

test = test %>%
  mutate(predictions_quad = predict(model_QUAD, test))
test %>% 
  summarize(MSE_quad = mean((mpg-predictions_quad)^2))

model_CUBIC = lm(mpg~poly(horsepower,3),data=train)

test = test %>%
  mutate(predictions_cubic = predict(model_CUBIC, test))
test %>%
  summarize(MSE_cubic = mean((mpg-predictions_cubic)^2))

MSE_quad
1 19.82259

MSE_cubic
1 19.78252

These error rates are 19.82 and 19.78, respectively. If we choose a different
training set instead, then we will obtain somewhat different errors on the
validation set. We can test this out by setting a different random seed:

In [7]:
set.seed(2)

train = Auto %>%
  sample_n(196)

test = Auto %>%
  setdiff(train)

model_LR2 = lm(mpg~horsepower, data=train)

test = test %>%
  mutate(predictions_slr = predict(model_LR2, test))

model_QUAD2=lm(mpg~poly(horsepower,2),data=train)

test = test %>%
  mutate(predictions_quad = predict(model_QUAD2, test))

model_CUBIC2=lm(mpg~poly(horsepower,3),data=train)

test = test %>%
  mutate(predictions_cubic =  predict(model_CUBIC2, test))

test %>%
  summarize(MSE_slr = mean((mpg-predictions_slr)^2),
            MSE_quad = mean((mpg - predictions_quad)^2),
            MSE_cube = mean((mpg - predictions_cubic)^2))

MSE_slr  MSE_quad MSE_cube
1 23.29559 18.90124 19.2574

Using this split of the observations into a training set and a validation
set, we find that the validation set error rates for the models with linear,
quadratic, and cubic terms are 23.30, 18.90, and 19.26, respectively.

These results are consistent with our previous findings: a model that
predicts `mpg` using a quadratic function of `horsepower` performs better than
a model that involves only a linear function of `horsepower`, and there is
little evidence in favor of a model that uses a cubic function of `horsepower`.

# 5.3.2 Leave-One-Out Cross-Validation

The LOOCV estimate can be automatically computed for any generalized
linear model using the `glm()` and `cv.glm()` functions. In the lab for Chapter 4, we used the `glm()` function to perform logistic regression by passing
in the `family="binomial"` argument. But if we use `glm()` to fit a model
without passing in the family argument, then it performs linear regression,
just like the `lm()` function. The following should yield identical models:

In [8]:
model_GLR = glm(mpg~horsepower, data=Auto)
coef(model_GLR)

model_LR = lm(mpg~horsepower, data=Auto)
coef(model_LR)

(Intercept)  horsepower 
 39.9358610  -0.1578447

(Intercept)  horsepower 
 39.9358610  -0.1578447

In this lab, we will perform linear
regression using the `glm()` function rather than the `lm()` function because
the latter can be used together with `cv.glm()` to perform cross-validation. The `cv.glm()` function is part of the `boot` library.

In [9]:
library(boot)
model_GLR = glm(mpg~horsepower, data=Auto)

cv_error = cv.glm(Auto, model_GLR)
cv_error$delta

[1] 24.23151 24.23114

The `cv.glm()` function produces a list with several components. The two
numbers in the delta vector contain the cross-validation results. In this case the numbers are identical (up to two decimal places) and correspond
to the LOOCV statistic: our cross-validation estimate for the test
error is approximately 24.23. Below, we'll discuss a situation in
which the two numbers differ. 

We can repeat this procedure for increasingly complex polynomial fits.
To automate the process, we use the `for()` function to initiate a for loop
which iteratively fits polynomial regressions for polynomials of order `i = 1`
to `i = 5` and computes the associated cross-validation error. 

This command may take a couple of minutes to run.

In [10]:
deltas = data.frame(delta1=0, delta2=0)
for (i in 1:5){
 model_GLR = glm(mpg~poly(horsepower,i), data=Auto)
 deltas[i,] = cv.glm(Auto, model_GLR)$delta
}
deltas

delta1   delta2  
1 24.23151 24.23114
2 19.24821 19.24787
3 19.33498 19.33448
4 19.42443 19.42371
5 19.03321 19.03242

Here we see a sharp drop in the estimated test MSE between
the linear and quadratic fits, but then no clear improvement from using
higher-order polynomials.

# 5.3.3 k-Fold Cross-Validation

The `cv.glm()` function can also be used to implement `k`-fold CV. Below we
use `k = 10`, a common choice for `k`, on the `Auto` data set. We once again set
a random seed and initialize a vector in which we will store the CV errors
corresponding to the polynomial fits of orders one to ten.

In [11]:
set.seed(1)
cv_errors = data.frame(delta1 = 0, delta2 = 0)

for (i in 1:10){
 model_GLR = glm(mpg~poly(horsepower,i), data=Auto)
 cv_errors[i, ] = cv.glm(Auto, model_GLR, K=10)$delta
}
cv_errors

delta1   delta2  
1  24.10716 24.09865
2  19.24186 19.22821
3  19.29106 19.27265
4  19.45578 19.42426
5  19.25644 19.21207
6  18.86037 18.82663
7  18.84461 18.80370
8  18.77549 18.73721
9  19.66007 19.55772
10 19.54238 19.43528

Notice that the computation time is **much** shorter than that of LOOCV.
(In principle, the computation time for LOOCV for a least squares linear
model should be faster than for `k`-fold CV, due to the availability of the
formula (5.2) for LOOCV; however, unfortunately the `cv.glm()` function
does not make use of this formula.) We still see little evidence that using
cubic or higher-order polynomial terms leads to lower test error than simply
using a quadratic fit.

We saw in Section 5.3.2 that the two numbers associated with delta are
essentially the same when LOOCV is performed. When we instead perform
`k`-fold CV, then the two numbers associated with delta differ slightly. The first is the standard `k`-fold CV estimate, as in (5.3). The second is a bias-corrected
version. On this data set, the two estimates are very similar to
each other.

# An Application to Default Data

Now that you're armed with more useful technique for resampling your data, let's try fitting a model for the `Default` dataset:

In [12]:
summary(Default)

 default    student       balance           income     
 No :9667   No :7056   Min.   :   0.0   Min.   :  772  
 Yes: 333   Yes:2944   1st Qu.: 481.7   1st Qu.:21340  
                       Median : 823.6   Median :34553  
                       Mean   : 835.4   Mean   :33517  
                       3rd Qu.:1166.3   3rd Qu.:43808  
                       Max.   :2654.3   Max.   :73554  

First we'll try just holding out a random 20% of the data:

In [13]:
for (i in 1:10){

    set.seed(i)

    train = Default %>%
      sample_frac(0.2)

    test = Default %>%
      setdiff(train)

    # Fit a logistic regression to predict default using balance
    model_LOGISTIC = glm(default~balance+student, data=train, family=binomial)

    # Use the model to predict the response on the test data
    glm_probs = data.frame(probs = predict(model_LOGISTIC, newdata=test, type="response"))

    # Confusion matrix
    glm_pred = glm_probs %>%
      mutate(pred = ifelse(probs>.5, "Yes", "No"))

    glm_pred = cbind(test, glm_pred)

    result = glm_pred %>%
      summarize(score = mean(pred == default))
    
    print(result)
}

  score
1 0.972
     score
1 0.972375
     score
1 0.974375
    score
1 0.97225
     score
1 0.973375
    score
1 0.97225
     score
1 0.974375
     score
1 0.973875
     score
1 0.973875
    score
1 0.97125


Our accuracy is really high on this data, but we're getting different error rates depending on how we choose our test set. That's no good!

Unfortunately this dataset is too big for us to run LOOCV, so we'll have to settle for `k`-fold. In the space below, build a logistic model on the full `Default`dataset and then run 5-fold cross-validation to get a more accurate estimate of your test error rate:

In [14]:
# Your code here

# 5.3.4 The Bootstrap

We illustrate the use of the bootstrap in the simple example of Section 5.2,
as well as on an example involving estimating the accuracy of the linear
regression model on the `Auto`data set.

# Estimating the Accuracy of a Statistic of Interest

One of the great advantages of the bootstrap approach is that it can be
applied in almost all situations. No complicated mathematical calculations
are required. Performing a bootstrap analysis in R entails only two steps.

- First, we must create a function that computes the statistic of interest.
- Second, we use the `boot()` function, which is part of the boot library, to perform the bootstrap by repeatedly sampling observations from the data set with replacement.

The `Portfolio`data set in the `ISLR` package is described in Section 5.2. It has variables called `X` and `Y`. 
To illustrate the use of the bootstrap on this data, we must first create
a function, `fn_alpha()`, which takes as input the data and outputs the estimate for $\alpha$ (described in more detail on page 187).

In [15]:
fn_alpha = function(data){
  
  # resample class needs to be "reconstituted" into a dataframe
  data <- as.data.frame(data)

  data %>%
  summarize(alpha = (var(Y)-cov(X,Y))/(var(X)+var(Y)-2*cov(X,Y))) %>%
  .$alpha
}

This function returns, or outputs, an estimate for $\alpha$ based on applying
(5.7) to the observations indexed by the argument index. For instance, the
following command tells `R` to estimate $\alpha$ using all 100 observations.

In [16]:
Portfolio %>%
  slice(1:100) %>%
  fn_alpha()

[1] 0.5758321

The next command uses the `sample()` function to randomly select 100 observations
from the range 1 to 100, with replacement. This is equivalent
to constructing a new bootstrap data set and recomputing $\hat{\alpha}$ based on the
new data set.

In [17]:
set.seed(1)
Portfolio %>%
  sample_n(100, replace=TRUE) %>%
  fn_alpha()

[1] 0.5963833

We can implement a bootstrap analysis by performing this command many
times, recording all of the corresponding estimates for $\alpha$, and computing the resulting standard deviation. However, the `bootstrap()`function automates
this approach. Below we produce $R = 1,000$ bootstrap estimates for $\alpha$:

In [18]:
bootstrap_Portfolio = Portfolio %>%
  modelr::bootstrap(1000) 

bootstrap_Portfolio %>%
  mutate(alpha = map_dbl(strap, fn_alpha)) 

strap                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                
1    -0.89525089, -1.56245433, -0.41708988, 1.04435573, -0.31556841, -1.73712385, 1.96641316, 2.15286790, -0.08120803, -0.89178179, -0.29320170, 0.50577917, 0.52675125, 1.06646932, 0.294

The final output shows that using the original data, $\hat{\alpha} = 0.5758$, and that the bootstrap estimate for $SE(\hat{\alpha})$ is 0.0886.

# Estimating the Accuracy of a Linear Regression Model

The bootstrap approach can be used to assess the variability of the coefficient
estimates and predictions from a statistical learning method. Here
we use the bootstrap approach in order to assess the variability of the
estimates for $\beta_0$ and $\beta_1$, the intercept and slope terms for the linear regression
model that uses horsepower to predict mpg in the Auto data set. We
will compare the estimates obtained using the bootstrap to those obtained
using the formulas for $SE(\hat{\beta}_0)$ and $SE(\hat{\beta}_1)$ described in Section 3.1.2.

We first create a simple function, `fn_model()`, which takes in the `Auto` data
and returns the intercept
and slope estimates for the linear regression model. We then apply this
function to the full set of 392 observations in order to compute the estimates
of $\beta_0$ and $\beta_1$ on the entire data set using the usual linear regression
coefficient estimate formulas from Chapter 3.

In [19]:
fn_model = function(data){
  lm(mpg ~ horsepower, data = data)
}

Auto %>% fn_model()


Call:
lm(formula = mpg ~ horsepower, data = data)

Coefficients:
(Intercept)   horsepower  
    39.9359      -0.1578  


The `fn_model()`function can also be used in order to create bootstrap estimates
for the intercept and slope terms by randomly sampling from among
the observations with replacement. Here we give two examples:

In [20]:
set.seed(1)
Auto_bootstrap = Auto %>%
  modelr::bootstrap(1)

df_regression_bootstrap_model = Auto_bootstrap %>%
  mutate(model = map(strap, fn_model)) 

df_regression_bootstrap_param = df_regression_bootstrap_model %>%
  mutate(param = map(model, tidy)) %>%
  select(.id, param) %>%
  unnest() %>%
  print()

Auto %>%
  modelr::bootstrap(1) %>%
  mutate(model = map(strap, fn_model))  %>%
  mutate(param = map(model, tidy)) %>%
  select(.id, param) %>%
  unnest() %>%
  print()

# A tibble: 2 x 6
  .id   term        estimate std.error statistic   p.value
  <chr> <chr>          <dbl>     <dbl>     <dbl>     <dbl>
1 1     (Intercept)   38.7     0.698        55.5 2.75e-187
2 1     horsepower    -0.148   0.00618     -24.0 8.88e- 79
# A tibble: 2 x 6
  .id   term        estimate std.error statistic   p.value
  <chr> <chr>          <dbl>     <dbl>     <dbl>     <dbl>
1 1     (Intercept)   40.0     0.697        57.4 2.24e-192
2 1     horsepower    -0.160   0.00637     -25.0 3.44e- 83


Next, we use the `bootstrap()` function to compute the standard errors of 1,000
bootstrap estimates for the intercept and slope terms:

In [21]:
set.seed(1)
Auto_bootstrap = Auto %>%
  modelr::bootstrap(1000)

df_regression_bootstrap_model = Auto_bootstrap %>%
  mutate(model = map(strap, fn_model)) 

df_regression_bootstrap_param = df_regression_bootstrap_model %>%
  mutate(param = map(model, tidy)) %>%
  select(.id, param) %>%
  unnest() %>%
  print()

df_regression_bootstrap_param %>%
  group_by(term) %>%
  summarize(point_est = mean(estimate),
            std_error = mean(std.error))

# A tibble: 2,000 x 6
   .id   term        estimate std.error statistic   p.value
   <chr> <chr>          <dbl>     <dbl>     <dbl>     <dbl>
 1 0001  (Intercept)   38.7     0.698        55.5 2.75e-187
 2 0001  horsepower    -0.148   0.00618     -24.0 8.88e- 79
 3 0002  (Intercept)   40.0     0.697        57.4 2.24e-192
 4 0002  horsepower    -0.160   0.00637     -25.0 3.44e- 83
 5 0003  (Intercept)   39.5     0.692        57.1 1.98e-191
 6 0003  horsepower    -0.156   0.00615     -25.3 2.27e- 84
 7 0004  (Intercept)   39.2     0.710        55.1 3.16e-186
 8 0004  horsepower    -0.150   0.00631     -23.8 7.92e- 78
 9 0005  (Intercept)   40.3     0.693        58.1 4.78e-194
10 0005  horsepower    -0.159   0.00620     -25.6 1.30e- 85
# … with 1,990 more rows


term        point_est  std_error  
1 (Intercept) 39.9607047 0.716724376
2 horsepower  -0.1581086 0.006445385

This indicates that the bootstrap estimate for $SE(\hat\beta_0)$ is 0.716, and that
the bootstrap estimate for $SE(\hat\beta_1)$ is 0.0064. As discussed in Section 3.1.2,
standard formulas can be used to compute the standard errors for the
regression coefficients in a linear model. These can be obtained using the
`summary()`function:

In [22]:
summary(lm(mpg~horsepower, data=Auto))


Call:
lm(formula = mpg ~ horsepower, data = Auto)

Residuals:
     Min       1Q   Median       3Q      Max 
-13.5710  -3.2592  -0.3435   2.7630  16.9240 

Coefficients:
             Estimate Std. Error t value Pr(>|t|)    
(Intercept) 39.935861   0.717499   55.66   <2e-16 ***
horsepower  -0.157845   0.006446  -24.49   <2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 4.906 on 390 degrees of freedom
Multiple R-squared:  0.6059,	Adjusted R-squared:  0.6049 
F-statistic: 599.7 on 1 and 390 DF,  p-value: < 2.2e-16


Note that the standard error estimates produced by the `summary()`function were a little different from the estimates obtained using the bootstrap. Does this indicate a problem with the bootstrap? In fact, it's just the opposite!

Recall that we found previously that the relationship between `horsepower` and `mpg` is better characterized by a quadratic model. Let's see how the error rates compare when we fit that instead of a linear model:

In [23]:
fn_boot = function(data){
  lm(mpg~horsepower + I(horsepower^2) ,data=data)
}

set.seed(1)
Auto_bootstrap = Auto %>%
  modelr::bootstrap(1000)

df_regression_bootstrap_model = Auto_bootstrap %>%
  mutate(model = map(strap, fn_boot)) 

df_regression_bootstrap_param = df_regression_bootstrap_model %>%
  mutate(param = map(model, tidy)) %>%
  select(.id, param) %>%
  unnest() %>%
  print()

df_regression_bootstrap_param %>%
  group_by(term) %>%
  summarize(point_est = mean(estimate),
            std_error = mean(std.error))

# A tibble: 3,000 x 6
   .id   term             estimate std.error statistic   p.value
   <chr> <chr>               <dbl>     <dbl>     <dbl>     <dbl>
 1 0001  (Intercept)      54.0      1.82         29.7  4.59e-102
 2 0001  horsepower       -0.423    0.0312      -13.6  1.24e- 34
 3 0001  I(horsepower^2)   0.00107  0.000120      8.97 1.26e- 17
 4 0002  (Intercept)      54.5      1.77         30.9  1.43e-106
 5 0002  horsepower       -0.426    0.0309      -13.8  1.48e- 35
 6 0002  I(horsepower^2)   0.00108  0.000122      8.79 4.77e- 17
 7 0003  (Intercept)      56.2      1.72         32.7  8.53e-114
 8 0003  horsepower       -0.457    0.0294      -15.5  1.13e- 42
 9 0003  I(horsepower^2)   0.00119  0.000114     10.4  1.46e- 22
10 0004  (Intercept)      53.0      1.84         28.8  2.37e- 98
# … with 2,990 more rows


term            point_est    std_error   
1 (Intercept)     56.916368760 1.8030368615
2 horsepower      -0.466574438 0.0312309253
3 I(horsepower^2)  0.001232793 0.0001228167

In [24]:
summary(lm(mpg~horsepower + I(horsepower^2), data=Auto))


Call:
lm(formula = mpg ~ horsepower + I(horsepower^2), data = Auto)

Residuals:
     Min       1Q   Median       3Q      Max 
-14.7135  -2.5943  -0.0859   2.2868  15.8961 

Coefficients:
                  Estimate Std. Error t value Pr(>|t|)    
(Intercept)     56.9000997  1.8004268   31.60   <2e-16 ***
horsepower      -0.4661896  0.0311246  -14.98   <2e-16 ***
I(horsepower^2)  0.0012305  0.0001221   10.08   <2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 4.374 on 389 degrees of freedom
Multiple R-squared:  0.6876,	Adjusted R-squared:  0.686 
F-statistic:   428 on 2 and 389 DF,  p-value: < 2.2e-16


Since this model provides a good fit to the data, there is now a better correspondence between the bootstrap estimates and the standard estimates of $SE(\hat\beta_0), SE(\hat\beta_1)$ and $SE(\hat\beta_2)$.

To get credit for this lab, please post your answers to the following questions:
- How did the cross-validated error rate compare to the models where you held out a validation set? Why do you think that is?
- How do you explain the discrepancy between the bootstrap evaluation and the standard error evaluation of the linear model predicting `mpg `from `horsepower`?
- What was the most confusing part of today's class?